### PR TITLE
Spotlight search labs

### PR DESCRIPTION
**Pull request by t3chguy**
Requires https://github.com/vector-im/element-web/pull/19699
Fixes https://github.com/vector-im/element-web/issues/19530

---
Here's what your changelog entry will look like:

## ✨ Features
 * Spotlight search labs ([\#7116](https://github.com/matrix-org/matrix-react-sdk/pull/7116)). Fixes vector-im/element-web#19530.

Preview: https://61b33e18c7cf7e0d5949a597--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

**Comment by niquewoodhouse** [comment]
@nadonomy @t3chguy 

Following last week's discussions, I've taken a look and propose this for first version to test:

Notes:
- Include in results names of Spaces and subspaces you've joined. 
  - Discussed with @t3chguy and seems trivial to include but would help users switching to other spaces, which would be one of the aims of a spotlight experience 
- Include a way to search for public rooms and a hint on how to search for messages

https://user-images.githubusercontent.com/1587166/144854913-2b2f8ed0-defa-4eca-bb4c-6a77b898fd25.mp4

**Review by niquewoodhouse — COMMENTED** [review]
**Recently viewed**

- [ ] Tooltip or alt text showing full name of room on hover
- [ ] For some reason the title doesn't seem to be centering to the avatar
- [ ] Not sure if this is easily possible but I think hiding the currently viewed room/space might be helpful because it makes space for another item in the list, and its currently showing so we could assume the user doesn't intend to be redirected to it

**General**

- [ ] Change `search for anything` to `search` as hint
- [ ] Seem to have lost the up and down to scroll results hint

**Comment by DoM1niC** [comment]
grafik
Status in Avatar drift away

**Comment by t3chguy** [comment]
image

@DoM1niC cannot reproduce, are you sure you're running the very latest version? that issue got fixed at some point.

**Review by niquewoodhouse — CHANGES_REQUESTED** [review]
**Recently viewed** 

- [x] Add 4px margin-bottom to the avatar to move room name slightly away from avatar

**Sections**

- [x] Add margin-bottom 8px to h4 titles, so the title is a bit closer to the contents of the section

**Comment by niquewoodhouse** [comment]
@nadonomy this is the spotlight search issue.  In a recent sync we discussed seeing if there's any way we could help people looking to start a DM/look for someone, as we saw in those unmoderated research sessions when someone has to find someone, they look to search/filter.  

My quickest suggestion would be to add a "People" item to the "Use X to search" section when searching, which could open the existing start chat dialog.  @t3chguy does that feel like a fairly trivial thing to include in this?

https://user-images.githubusercontent.com/1587166/145044599-df8506a8-e14a-4799-9488-38524c0b4307.mp4

**Comment by t3chguy** [comment]
> @t3chguy does that feel like a fairly trivial thing to include in this?

Yup!

**Comment by t3chguy** [comment]
image

Done

**Review by niquewoodhouse — APPROVED** [review]

File: res/css/views/dialogs/_SpotlightDialog.scss [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+    height: 60%;
+    padding: 0;
+    contain: unset; // needed for .mx_SpotlightDialog_keyboardPrompt to not be culled
+
+    .mx_SpotlightDialog_keyboardPrompt {
+        position: absolute;
+        padding: 8px;
+        border-radius: 8px;
+        background-color: $background;
+        top: -60px;
```
Comment (turt2live) [inline res/css/views/dialogs/_SpotlightDialog.scss:30]: 60px is relative to what?

File: src/components/views/dialogs/SpotlightDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+import DecoratedRoomAvatar from "../avatars/DecoratedRoomAvatar";
+import { Action } from "../../../dispatcher/actions";
+import Modal from "../../../Modal";
+import GenericFeatureFeedbackDialog from "./GenericFeatureFeedbackDialog";
+import AccessibleTooltipButton from "../elements/AccessibleTooltipButton";
+import RoomViewStore from "../../../stores/RoomViewStore";
+import { showStartChatInviteDialog } from "../../../RoomInvite";
+
+const MAX_RECENT_SEARCHES = 10;
+const SECTION_LIMIT = 50;
```
Comment (turt2live) [inline src/components/views/dialogs/SpotlightDialog.tsx:66]: 50 whats? (units, comment, or descriptive const please)

File: src/components/views/dialogs/SpotlightDialog.tsx [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+const RECENT_SEARCHES_LS_KEY = "mx_SpotlightDialog_recent_searches";
```
Comment (turt2live) [inline src/components/views/dialogs/SpotlightDialog.tsx:67]: arguably should be a setting, but won't fight for it to be one here

**Review by turt2live — COMMENTED** [review]
overall looks fine - just a couple nitpicks

File: src/components/views/dialogs/SpotlightDialog.tsx (hunk 3 above)
Comment (t3chguy) [reply src/components/views/dialogs/SpotlightDialog.tsx:67]: Good point, syncing it could be neat

**Review by turt2live — APPROVED** [review]
Thanks :)

**Pull request merged**